File: src/adapters/telegram/image_fetcher.ts
import type { FetchedImage, ImageFetcher, ImageReference } from './types.ts';

export class ImageTooLargeError extends Error {
  override name = 'ImageTooLargeError';
}
export class UnsupportedImageError extends Error {
  override name = 'UnsupportedImageError';
}

export interface TelegramFileURLer {
  getFileDirectURL(signal: AbortSignal, fileId: string): Promise<string>;
}

export class TelegramImageFetcher implements ImageFetcher {
  readonly #urler: TelegramFileURLer;
  readonly #maxBytes: number;
  readonly #fetcher: typeof fetch;

  constructor(urler: TelegramFileURLer, maxBytes: number, fetcher: typeof fetch = fetch) {
    if (maxBytes <= 0) throw new Error('telegram image size limit must be positive');
    this.#urler = urler;
    this.#maxBytes = maxBytes;
    this.#fetcher = fetcher;
  }

  async fetchImage(signal: AbortSignal, reference: ImageReference): Promise<FetchedImage> {
    if (!reference.fileId) throw new Error('telegram image file ID is required');
    if ((reference.declaredSize ?? 0) > this.#maxBytes) {
      throw new ImageTooLargeError('telegram image exceeds size limit');
    }
    const url = await this.#urler.getFileDirectURL(signal, reference.fileId);
    const response = await this.#fetcher(url, { signal });
    if (!response.ok) {
      throw new Error(`telegram image download returned HTTP status ${response.status}`);
    }
    if (Number(response.headers.get('content-length')) > this.#maxBytes) {
      throw new ImageTooLargeError('telegram image exceeds size limit');
    }
    if (!response.body) throw new Error('telegram image response has no body');

    const reader = response.body.getReader();
    const chunks: Uint8Array[] = [];
    let total = 0;
    try {
      while (true) {
        const next = await reader.read();
        if (next.done) break;
        total += next.value.byteLength;
        if (total > this.#maxBytes) {
          throw new ImageTooLargeError('telegram image exceeds size limit');
        }
        chunks.push(next.value);
      }
    } finally {
      reader.releaseLock();
    }
    const data = new Uint8Array(total);
    let offset = 0;
    for (const chunk of chunks) {
      data.set(chunk, offset);
      offset += chunk.byteLength;
    }
    const mimeType = detectImageMime(data);
    if (
      !mimeType || (reference.declaredMime && !mimeCompatible(reference.declaredMime, mimeType))
    ) {
      throw new UnsupportedImageError('unsupported image content');
    }
    return { mimeType, data };
  }
}

export function detectImageMime(data: Uint8Array): string | undefined {
  if (data.length >= 3 && data[0] === 0xff && data[1] === 0xd8 && data[2] === 0xff) {
    return 'image/jpeg';
  }
  if (
    data.length >= 8 && data[0] === 0x89 && data[1] === 0x50 && data[2] === 0x4e &&
    data[3] === 0x47 && data[4] === 0x0d && data[5] === 0x0a && data[6] === 0x1a && data[7] === 0x0a
  ) return 'image/png';
  if (
    data.length >= 12 && ascii(data.slice(0, 4)) === 'RIFF' && ascii(data.slice(8, 12)) === 'WEBP'
  ) return 'image/webp';
  return undefined;
}

function ascii(data: Uint8Array): string {
  return String.fromCharCode(...data);
}
function mimeCompatible(declared: string, detected: string): boolean {
  const normalized = declared.toLowerCase().split(';', 1)[0]!.trim();
  return normalized === detected;
}

File: telegram/image_fetcher.go
package telegram

import (
	"bytes"
	"context"
	"errors"
	"fmt"
	"io"
	"net/http"
)

var (
	ErrImageTooLarge    = errors.New("telegram image exceeds size limit")
	ErrUnsupportedImage = errors.New("unsupported image content")
)

type TelegramFileURLer interface {
	GetFileDirectURL(fileID string) (string, error)
}

type FetchedImage struct {
	MIMEType string
	Data     []byte
}

type ImageFetcher interface {
	FetchImage(ctx context.Context, ref ImageReference) (FetchedImage, error)
}

type TelegramImageFetcher struct {
	urler TelegramFileURLer
	http  interface {
		Do(*http.Request) (*http.Response, error)
	}
	maxBytes int64
}

func NewTelegramImageFetcherForBot(bot TelegramFileURLer, maxBytes int64) (*TelegramImageFetcher, error) {
	return NewTelegramImageFetcher(bot, newTelegramHTTPClient(), maxBytes)
}

func NewTelegramImageFetcher(urler TelegramFileURLer, httpClient interface {
	Do(*http.Request) (*http.Response, error)
}, maxBytes int64) (*TelegramImageFetcher, error) {
	if urler == nil {
		return nil, errors.New("telegram file URLer is required")
	}
	if httpClient == nil {
		return nil, errors.New("telegram image HTTP client is required")
	}
	if maxBytes <= 0 {
		return nil, errors.New("telegram image size limit must be positive")
	}
	return &TelegramImageFetcher{urler: urler, http: httpClient, maxBytes: maxBytes}, nil
}

func (f *TelegramImageFetcher) FetchImage(ctx context.Context, ref ImageReference) (FetchedImage, error) {
	if f == nil || f.urler == nil || f.http == nil {
		return FetchedImage{}, errors.New("telegram image fetcher unavailable")
	}
	if ref.FileID == "" {
		return FetchedImage{}, errors.New("telegram image file ID is required")
	}
	if ref.DeclaredSize > f.maxBytes {
		return FetchedImage{}, ErrImageTooLarge
	}
	url, err := f.urler.GetFileDirectURL(ref.FileID)
	if err != nil {
		return FetchedImage{}, errors.New("resolve telegram image failed")
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return FetchedImage{}, errors.New("create telegram image download request failed")
	}
	resp, err := f.http.Do(req)
	if err != nil {
		return FetchedImage{}, errors.New("download telegram image failed")
	}
	defer resp.Body.Close()
	if resp.StatusCode < http.StatusOK || resp.StatusCode >= http.StatusMultipleChoices {
		return FetchedImage{}, fmt.Errorf("telegram image download returned HTTP status %d", resp.StatusCode)
	}
	data, err := io.ReadAll(io.LimitReader(resp.Body, f.maxBytes+1))
	if err != nil {
		return FetchedImage{}, errors.New("read telegram image failed")
	}
	if int64(len(data)) > f.maxBytes {
		return FetchedImage{}, ErrImageTooLarge
	}
	mimeType, ok := DetectImageMIME(data)
	if !ok {
		return FetchedImage{}, ErrUnsupportedImage
	}
	return FetchedImage{MIMEType: mimeType, Data: data}, nil
}

func DetectImageMIME(data []byte) (string, bool) {
	switch http.DetectContentType(data) {
	case "image/jpeg":
		return "image/jpeg", true
	case "image/png":
		return "image/png", true
	}
	if len(data) >= 12 && bytes.Equal(data[:4], []byte("RIFF")) && bytes.Equal(data[8:12], []byte("WEBP")) {
		return "image/webp", true
	}
	return "", false
}
